# House Prices
#### Ryan Peralta, Jonathan Wang, Yillian Wu
#### CS 4661
#### Fall 2019
https://www.kaggle.com/c/house-prices-advanced-regression-techniques/overview



In [1]:
# Importing numpy and pandas and other modules
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn import metrics
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn import preprocessing
import matplotlib.pyplot as plt

### Reading the csv into a dataframe

In [2]:
#Reading the csv downloaded from Kaggle into a dataframe
houses_df = pd.read_csv("House_Prices.csv")

#Combined the seperated csv from kaggle into one csv for better testing.
houses_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal   208500.0  
1   2007        WD         Normal   181500.0  
2   2008        WD         Normal   223500.0  
3   2006        WD        Abnorml   140000.0  
4   2008        WD         Normal   250000.0  

[5 rows x 81 columns]

## 1.) Linear Regression

### Initial run with only 10 selected features
Establish a baseline for the project.
Every model we use after this should have a lower RMSE.

In [3]:
# Creating the Feature Matrix:
# hand selected features that we thought would give us a decent regression.
feature_cols = ['LotArea','Neighborhood', 'BldgType', 'HouseStyle','YearBuilt', 'FullBath', 'YearRemodAdd', 'Foundation','SaleType', 'SaleCondition']

# Get X from the training data
X = houses_df[feature_cols]
# select the target (last column) from the DataFrame
y = houses_df['SalePrice']
X.head()

LotArea Neighborhood BldgType HouseStyle  YearBuilt  FullBath  \
0     8450      CollgCr     1Fam     2Story       2003         2   
1     9600      Veenker     1Fam     1Story       1976         2   
2    11250      CollgCr     1Fam     2Story       2001         2   
3     9550      Crawfor     1Fam     2Story       1915         1   
4    14260      NoRidge     1Fam     2Story       2000         2   

   YearRemodAdd Foundation SaleType SaleCondition  
0          2003      PConc       WD        Normal  
1          1976     CBlock       WD        Normal  
2          2002      PConc       WD        Normal  
3          1970     BrkTil       WD       Abnorml  
4          2000      PConc       WD        Normal

### Use label encoding to convert the string values into numbers

In [4]:
# label_encoder object knows how to understand word labels. 
label_encoder = preprocessing.LabelEncoder() 
  
# Encode labels in column 'species'. 
X['Neighborhood']= label_encoder.fit_transform(X['Neighborhood']) 
X['BldgType']= label_encoder.fit_transform(X['BldgType']) 
X['HouseStyle']= label_encoder.fit_transform(X['HouseStyle']) 
X['Foundation']= label_encoder.fit_transform(X['Foundation'])
X['SaleType'] = label_encoder.fit_transform(X['SaleType'].astype(str))
X['SaleCondition']= label_encoder.fit_transform(X['SaleCondition'])  
X.head()


C:\Users\rpera\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """
C:\Users\rpera\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\rpera\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#

LotArea  Neighborhood  BldgType  HouseStyle  YearBuilt  FullBath  \
0     8450             5         0           5       2003         2   
1     9600            24         0           2       1976         2   
2    11250             5         0           5       2001         2   
3     9550             6         0           5       1915         1   
4    14260            15         0           5       2000         2   

   YearRemodAdd  Foundation  SaleType  SaleCondition  
0          2003           2         8              4  
1          1976           1         8              4  
2          2002           2         8              4  
3          1970           0         8              0  
4          2000           2         8              4

### Splitting the Dataset into testing and training

In [21]:
# Splitting the dataset into testing and training:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=2)

In [22]:
linear_reg_houses = LinearRegression()
# fitting the model to the training data:
linear_reg_houses.fit(X_train, y_train)

# printing Theta0 using attribute "intercept_":
print(linear_reg_houses.intercept_)

# printing [Theta1, Theta2, Theta3] using attribute "coef_":
print(linear_reg_houses.coef_)

10955666870105.234
[-2.28045775e+09 -7.61549092e+12  1.87370774e+12 -2.60862944e+11
  1.12687388e+12 -1.44771928e+12 -7.95367733e+11 -7.92984898e+11
 -2.37457946e+12 -1.94532776e+12 -1.48271482e+12 -1.49348563e+12
 -7.17318799e+12 -5.80356203e+11 -1.02037871e+12  4.42405342e+11
  1.55223263e+11 -2.84799240e+11 -4.40022480e+11 -4.40022495e+11
 -4.40022498e+11 -4.40022474e+11 -4.40022486e+11 -4.40022502e+11
 -4.40022494e+11 -4.40022478e+11 -4.40022494e+11 -4.40022421e+11
 -4.40022452e+11 -4.40022494e+11 -4.40022498e+11 -4.40022502e+11
 -4.40022499e+11 -4.40022492e+11 -4.40022438e+11 -4.40022502e+11
 -4.40022469e+11  9.24490227e+06  9.23178181e+06  9.23154905e+06
  9.20424755e+06  9.21106139e+06 -1.60150542e+08 -1.60178250e+08
 -1.60159292e+08 -1.60134470e+08 -1.60129262e+08 -1.60148502e+08
 -1.60158894e+08 -1.60155902e+08  2.83188245e+07  2.83186709e+07
  2.83233082e+07  2.82996896e+07  2.83323481e+07  2.83233471e+07
  3.32691142e+07  3.32875469e+07  3.33215958e+07  3.32784960e+07
  3.33

In [7]:
# make predictions on the testing set
y_prediction = linear_reg_houses.predict(X_test)

# Calculating "Mean Square Error" (MSE):
mse = metrics.mean_squared_error(y_test, y_prediction)

# Using numpy sqrt function to take the square root and calculate "Root Mean Square Error" (RMSE)
rmse = np.sqrt(mse)

print(rmse)

49367.798232765104


### RMSE = 49367.79
This is the baseling RMSE that we are going to be using.
The goal for the other techniques we are going to use is to lower the RMSE and see which model is the best.

In [28]:
#%matplotlib inline
# Plot outputs
#plt.figure()
#plt.scatter(X_test, y_test,  color='black')
#plt.plot(X_test, y_prediction, color='blue', linewidth=3)

#plt.xticks(())
#plt.yticks(())

#plt.show()

## 1a.) Linear Regression with OneHot encoding
This is going to build off of the last model, but after using label encoding, we are going to use onehot encoding as well.

In [10]:
#Reading the csv downloaded from Kaggle into a dataframe
houses_df = pd.read_csv("House_Prices.csv")

#Combined the seperated csv from kaggle into one csv for better testing.

# Creating the Feature Matrix:
# hand selected features that we thought would give us a decent regression.
feature_cols = ['LotArea','Neighborhood', 'BldgType', 'HouseStyle','YearBuilt', 'FullBath', 'YearRemodAdd', 'Foundation','SaleType', 'SaleCondition']

# Get X from the training data
X = houses_df[feature_cols]
# select the target (last column) from the DataFrame
y = houses_df['SalePrice']

# label_encoder object knows how to understand word labels. 
label_encoder = preprocessing.LabelEncoder() 
  
# Encode labels in column 'species'. 
X['Neighborhood']= label_encoder.fit_transform(X['Neighborhood']) 
X['BldgType']= label_encoder.fit_transform(X['BldgType']) 
X['HouseStyle']= label_encoder.fit_transform(X['HouseStyle']) 
X['Foundation']= label_encoder.fit_transform(X['Foundation'])
X['SaleType'] = label_encoder.fit_transform(X['SaleType'].astype(str))
X['SaleCondition']= label_encoder.fit_transform(X['SaleCondition'])  
X.head()


C:\Users\rpera\Anaconda3\lib\site-packages\ipykernel_launcher.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\rpera\Anaconda3\lib\site-packages\ipykernel_launcher.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\rpera\Anaconda3\lib\site-packages\ipykernel_launcher.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexi

LotArea  Neighborhood  BldgType  HouseStyle  YearBuilt  FullBath  \
0     8450             5         0           5       2003         2   
1     9600            24         0           2       1976         2   
2    11250             5         0           5       2001         2   
3     9550             6         0           5       1915         1   
4    14260            15         0           5       2000         2   

   YearRemodAdd  Foundation  SaleType  SaleCondition  
0          2003           2         8              4  
1          1976           1         8              4  
2          2002           2         8              4  
3          1970           0         8              0  
4          2000           2         8              4

In [13]:
from sklearn.preprocessing import OneHotEncoder
onehotencoder = OneHotEncoder(categorical_features = [1,2,3,7,8,9])
X_onehot = onehotencoder.fit_transform(X).toarray()
X = pd.DataFrame(X_onehot)
X.head()

C:\Users\rpera\Anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:451: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


0    1    2    3    4    5    6    7    8    9   ...   66   67   68   69  \
0  1.0  0.0  0.0  1.0  1.0  0.0  1.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0   
1  1.0  0.0  0.0  1.0  1.0  0.0  1.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0   
2  1.0  0.0  0.0  1.0  1.0  0.0  1.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0   
3  1.0  0.0  0.0  1.0  1.0  0.0  1.0  0.0  0.0  1.0  ...  1.0  0.0  0.0  0.0   
4  1.0  0.0  0.0  1.0  1.0  0.0  1.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0   

    70   71       72      73   74      75  
0  1.0  0.0   8450.0  2003.0  2.0  2003.0  
1  1.0  0.0   9600.0  1976.0  2.0  1976.0  
2  1.0  0.0  11250.0  2001.0  2.0  2002.0  
3  0.0  0.0   9550.0  1915.0  1.0  1970.0  
4  1.0  0.0  14260.0  2000.0  2.0  2000.0  

[5 rows x 76 columns]

In [15]:
# Splitting the dataset into testing and training:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=2)

linear_reg_houses = LinearRegression()
# fitting the model to the training data:
linear_reg_houses.fit(X_train, y_train)

# printing Theta0 using attribute "intercept_":
print(linear_reg_houses.intercept_)

# printing [Theta1, Theta2, Theta3] using attribute "coef_":
print(linear_reg_houses.coef_)

# make predictions on the testing set
y_prediction = linear_reg_houses.predict(X_test)

10955666870105.234
[-2.28045775e+09 -7.61549092e+12  1.87370774e+12 -2.60862944e+11
  1.12687388e+12 -1.44771928e+12 -7.95367733e+11 -7.92984898e+11
 -2.37457946e+12 -1.94532776e+12 -1.48271482e+12 -1.49348563e+12
 -7.17318799e+12 -5.80356203e+11 -1.02037871e+12  4.42405342e+11
  1.55223263e+11 -2.84799240e+11 -4.40022480e+11 -4.40022495e+11
 -4.40022498e+11 -4.40022474e+11 -4.40022486e+11 -4.40022502e+11
 -4.40022494e+11 -4.40022478e+11 -4.40022494e+11 -4.40022421e+11
 -4.40022452e+11 -4.40022494e+11 -4.40022498e+11 -4.40022502e+11
 -4.40022499e+11 -4.40022492e+11 -4.40022438e+11 -4.40022502e+11
 -4.40022469e+11  9.24490227e+06  9.23178181e+06  9.23154905e+06
  9.20424755e+06  9.21106139e+06 -1.60150542e+08 -1.60178250e+08
 -1.60159292e+08 -1.60134470e+08 -1.60129262e+08 -1.60148502e+08
 -1.60158894e+08 -1.60155902e+08  2.83188245e+07  2.83186709e+07
  2.83233082e+07  2.82996896e+07  2.83323481e+07  2.83233471e+07
  3.32691142e+07  3.32875469e+07  3.33215958e+07  3.32784960e+07
  3.33

In [16]:
# Calculating "Mean Square Error" (MSE):
mse = metrics.mean_squared_error(y_test, y_prediction)

# Using numpy sqrt function to take the square root and calculate "Root Mean Square Error" (RMSE)
rmse = np.sqrt(mse)

print(rmse)

47786.45297361398


### RMSE = 47786.45
This is the new lowest RMSE.
One hot encoding on the previous model allowed us to get a lower value.

We should look into using label and one hot encoding for the other regression techniques we are going to use.

## 1b.) Linear Regression and cross validation
Continue using one hot encoding but now use cross validation to get a more accuracte RMSE

In [29]:
#Reading the csv downloaded from Kaggle into a dataframe
houses_df = pd.read_csv("House_Prices.csv")

#Combined the seperated csv from kaggle into one csv for better testing.

# Creating the Feature Matrix:
# hand selected features that we thought would give us a decent regression.
feature_cols = ['LotArea','Neighborhood', 'BldgType', 'HouseStyle','YearBuilt', 'FullBath', 'YearRemodAdd', 'Foundation','SaleType', 'SaleCondition']

# Get X from the training data
X = houses_df[feature_cols]
# select the target (last column) from the DataFrame
y = houses_df['SalePrice']

# label_encoder object knows how to understand word labels. 
label_encoder = preprocessing.LabelEncoder() 
  
# Encode labels in column 'species'. 
X['Neighborhood']= label_encoder.fit_transform(X['Neighborhood']) 
X['BldgType']= label_encoder.fit_transform(X['BldgType']) 
X['HouseStyle']= label_encoder.fit_transform(X['HouseStyle']) 
X['Foundation']= label_encoder.fit_transform(X['Foundation'])
X['SaleType'] = label_encoder.fit_transform(X['SaleType'].astype(str))
X['SaleCondition']= label_encoder.fit_transform(X['SaleCondition'])  

C:\Users\rpera\Anaconda3\lib\site-packages\ipykernel_launcher.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\rpera\Anaconda3\lib\site-packages\ipykernel_launcher.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\rpera\Anaconda3\lib\site-packages\ipykernel_launcher.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexi

In [48]:
# function cross_val_score performs Cross Validation:
linear_reg_houses = LinearRegression()
accuracy_list_lr = cross_val_score(linear_reg_houses, X, y, cv=10, scoring='neg_mean_squared_error')

In [69]:
def List_Average(x): 
    return sum(x) / len(x) 

rmse_list = []
for l in accuracy_list_lr:
    x = np.sqrt(-1* l)
    #print(x)
    rmse_list.append(x)


In [71]:
#for a in rmse_list:
    #print(a)
    
print("Linear Regression:                   ", List_Average(rmse_list))

Linear Regression:                    46602.33014125868


### RMSE = 46602.33
This is the new lowest RMSE.
By using cross validation, we were able to lower the RMSE from part 1a
